awscli/alias.py

```diff
@@ -0,0 +1,206 @@
+        # the alias proxied to.
+        return self._command_table[parsed_alias_args.command](
+            remaining, parsed_globals)
+
+    def _get_alias_args(self):
+        try:
+            alias_args = shlex.split(self._alias_value)
+        except ValueError as e:
+            raise InvalidAliasException(
+                'Value of alias "%s" could not be parsed. '
+                'Received error: %s when parsing:\n%s' % (
+                    self._alias_name, e, self._alias_value)
+            )
+
+        alias_args = [arg.strip(os.linesep) for arg in alias_args]
+        LOG.debug(
+            'Expanded subcommand alias %r with value: %r to: %r',
+            self._alias_name, self._alias_value, alias_args
+        )
+        return alias_args
+
+    def _update_parsed_globals(self, parsed_alias_args, parsed_globals):
+        global_params_to_update = self._get_global_parameters_to_update(
+            parsed_alias_args)
+        # Emit the top level args parsed event to ensure all possible
+        # customizations that typically get applied are applied to the
+        # global parameters provided in the alias before updating
+        # the original provided global parameter values
+        # and passing those onto subsequent commands.
+        emit_top_level_args_parsed_event(self._session, parsed_alias_args)
+        for param_name in global_params_to_update:
+            updated_param_value = getattr(parsed_alias_args, param_name)
+            setattr(parsed_globals, param_name, updated_param_value)
+
+    def _get_global_parameters_to_update(self, parsed_alias_args):
+        # Retrieve a list of global parameters that the newly parsed args
+        # from the alias will have to clobber from the originally provided
+        # parsed globals.
+        global_params_to_update = []
+        for parsed_param, value in vars(parsed_alias_args).items():
+            # To determine which parameters in the alias were global values
+            # compare the parsed parameter to its default as specified by
+            # the parser. If the parsed value differs from the default value
+            # that global parameter must have been provided.
+            if self._parser.get_default(parsed_param) != value:
+                if parsed_param in self.UNSUPPORTED_GLOBAL_PARAMETERS:
+                    raise InvalidAliasException(
+                        'Global parameter "--%s" detected in alias "%s" '
+                        'which is not support in subcommand aliases.' % (
+                            parsed_param, self._alias_name))
+                else:
+                    global_params_to_update.append(parsed_param)
+        return global_params_to_update
+
+
+class ExternalAliasCommand(BaseAliasCommand):
+    """Command for `toplevel` external alias"""
+    def __call__(self, args, parsed_globals):
+        command_components = [
+            self._alias_value.lstrip('!')
```

Comment:
Is `lstrip()` what we want here?  Or do you only want to remove the first `!` char (`self._alias_value[1:]` if the first char is a `!`).